### PR TITLE
Support raw/binary output with UDPWriter

### PR DESCRIPTION
**Pull request by veggiemike**
Ahoy!  I ran into a couple issues this last cruise aboard the E/V Nautilus and decided to get my hands dirty and fix them.  My primary issue was that we couldn't read raw binary data with a `NetworkReader` and then send it unmodified to a couple different systems using `UDPWriter` instances.  `NetworkReader` handles raw/binary data fine by setting `encoding=None`, but `UDPWriter` was blindly encoding it into UTF-8, which obviously garbles it all up.

I tried to keep this branch focused on just this issue, but couldn't help but fix a couple other problems I bumped into along the way.  I actually have a whole 2nd branch full of `SerialWriter`, `NetworkWriter`, and testsuite fixes to send along after this, but figured I'd start small(ish).  We're planning on using OpenRVDAS more and more on the Nautilus, so I'm going to keep on find-and-fixing while I'm out here.  

All the tests that passed before still pass (there are some `SerialWriter` errors already present that I've got fixed in my next branch).

### CODE REVIEW

**Comment by davidpablocohn** [comment]
Hi Mike - this looks great, thank you!

I'll dive into reviewing shortly, but may I ask that you refile the pull
request to pull it into the OpenRVDAS dev branch? I try to aggregate
changes there, then push everything out from dev-> master at once.

**Comment by veggiemike** [comment]
No problem, Pablo!  I've changed the pull target to OceanDataTools:dev.  Thanks!

**Comment by davidpablocohn** [comment]
Excellent - Thanks! Getting right on this. (And yeah, the feature/bug fix
makes perfect sense - thank you for diving in on that!)

File: logger/writers/udp_writer.py [hunk 1]
```diff
@@ -50,28 +52,61 @@ def __init__(self, port, destination='',
...
+        if eol is not None and self.encoding:
+            eol = self._unescape_str(eol)
         self.eol = eol
 
         self.target_str = 'interface: %s, destination: %s, port: %d' % (
             interface, destination, port)
 
+        # do name resolution once in the constructor
+        #
+        # NOTE: This means the hostname must be valid when we start, otherwise
```
Comment (davidpablocohn) [inline logger/writers/udp_writer.py:87]: Good  documentation - thanks!

File: logger/writers/udp_writer.py [hunk 2]
```diff
@@ -95,20 +130,23 @@ def __init__(self, port, destination='',
...
         # If no destination, it's a broadcast; set flag allowing broadcast and
         # set dest to special string
-        else:
+        elif not destination:
             destination = '<broadcast>'
 
         self.destination = destination
-        self.port = port
+        # make sure port gets stored as an int, even if passed in as a string
+        self.port = int(port)
```
Comment (davidpablocohn) [inline logger/writers/udp_writer.py:142]: Good catch.

File: logger/writers/udp_writer.py [hunk 3]
```diff
@@ -175,18 +213,33 @@ def write(self, record):
...
+                bytes_sent = self.socket.send(self._encode_str(record))
 
                 # If here, write at least partially succeeded. Reset warnings
-                if self.num_warnings == self.warning_limit:
-                    logging.info('UDPWriter.write() succeeded in writing after series of '
-                                 'failures; resetting warnings.')
-                self.num_warnings = 0  # we've succeeded
+                #
+                # NOTE: If the host is unreachable, every other send will fail.
+                #       Since UDP doesn't actually know it failed, the initial
```
Comment (davidpablocohn) [inline logger/writers/udp_writer.py:223]: Excellent.

**Review by davidpablocohn — APPROVED** [review]
Very clean code - thanks for thise!

File: logger/writers/udp_writer.py [hunk 4]
```diff
@@ -10,18 +9,18 @@
...
 
-class UDPWriter(NetworkWriter):
+class UDPWriter(Writer):
     """Write UDP packets to network."""
 
     def __init__(self, port, destination='',
                  interface='',  # DEPRECATED!
-                 ttl=3, num_retry=2, warning_limit=5, eol=''):
-        """
-        Write text records to a network socket.
```
[the remark this answers is not in the document]
Comment (veggiemike) [reply logger/writers/udp_writer.py:24]: Good catch, I'll change that in my next batch of changes.

File: logger/writers/udp_writer.py [hunk 5]
```diff
@@ -50,28 +52,61 @@ def __init__(self, port, destination='',
...
+                https://docs.python.org/3/library/codecs.html#standard-encodings
+
+        encoding_errors - 'ignore' by default. Other error strategies are
+                'strict', 'replace', and 'backslashreplace', described here:
+                https://docs.python.org/3/howto/unicode.html#encodings
         ```
+
         """
-        self.ttl = ttl
+        super().__init__(input_format=Text,
```
[the remark this answers is not in the document]
Comment (veggiemike) [reply logger/writers/udp_writer.py:67]: Sounds good, I'll ditch that line and not "fix" it anywhere else it's missing.

**Comment by veggiemike** [comment]
Thanks, Pablo!

**Pull request merged**